prymer/api/picking.py

```diff
@@ -131,50 +133,89 @@ def build_primer_pairs(
+    left_primers = sorted(left_primers, key=lambda p: p.span.start)
+    right_primers = sorted(right_primers, key=lambda p: p.span.end)
+
     # Grab the sequence we'll use to fill in the amplicon sequence
     with FastaFile(f"{fasta_path}") as fasta:
         region_start = min(p.span.start for p in left_primers)
         region_end = max(p.span.end for p in right_primers)
         bases = fasta.fetch(target.refname, region_start - 1, region_end)
 
+    # Each tuple is left_idx, right_idx, penalty, tm
+    pairings: list[Tuple[int, int, float, float]] = []
+
+    # generate all the primer pairs that don't violate hard size and Tm constraints
+    first_right_primer_idx = 0
+
+    # Nested loops over indices are used here so that we can skip potentially large chunks of
+    # the cartesian product, based on the fact that we're sorting the left and right primers.
+    # Two things are relied upon:
+    #   1. If we encounter a left/right combo that either has the right primer leftward of the
+    #      left primer _or_ generates a too-short amplicon, the neither that right primer nor
+    #      any previous right primer can make a valid combination with any subsequent left primer.
+    #   2. If we encounter a left/right combo that generates a too-large amplicon, then no
+    #      subsequent right-primer can make a valid combination with that left primer
+    for i in range(0, len(left_primers)):
+        for j in range(first_right_primer_idx, len(right_primers)):
+            lp = left_primers[i]
+            rp = right_primers[j]
+
+            # If the right primer isn't "to the right" of the left primer, move on
+            if rp.span.start < lp.span.start or lp.span.end > rp.span.end:
+                first_right_primer_idx = max(first_right_primer_idx, j+1)
+                continue
+
+            amp_span = PrimerPair.calculate_amplicon_span(lp, rp)
+
+            if amp_span.length < amplicon_sizes.min:
+                first_right_primer_idx = max(first_right_primer_idx, j+1)
+                continue
+
+            if amp_span.length > amplicon_sizes.max:
+                break  # break in this case because all subsequent rps will yield longer amplicons
+
+            # Since the amplicon span and the region_start are both 1-based, the minuend
+            # becomes a zero-based offset
+            amp_bases = bases[amp_span.start - region_start : amp_span.end - region_start + 1]
+            amp_tm = calculate_long_seq_tm(amp_bases)
+
+            if amp_tm < amplicon_tms.min or amp_tm > amplicon_tms.max:
+                continue
+
+            penalty = score(
+                left_primer=lp,
+                right_primer=rp,
+                amplicon=amp_span,
+                amplicon_tm=amp_tm,
+                amplicon_sizes=amplicon_sizes,
+                amplicon_tms=amplicon_tms,
+                weights=weights,
+            )
+
+            pairings.append((i, j, penalty, amp_tm))
```

Comment:
**suggestion** You might consider constructing `PrimerPair` here, and then using `dataclasses.replace` to populate `bases` below (where the `PrimerPair` is currently being instantiated).

That way the function is using established data structures from the library instead of tuples, the sort can reference `pair.penalty` instead of a tuple element, and I think it would save a few LOC overall.

Take it or leave it though 😄